### PR TITLE
Backend/feat/unified-pool

### PR DESCRIPTION
**Pull request by enigma-commits**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DriverPool.hs [hunk 1]
```diff
@@ -635,7 +635,8 @@ filterOutGoHomeDriversAccordingToHomeLocation randomDriverPool CalculateGoHomeDr
           keepHiddenForSeconds = Seconds 0,
           goHomeReqId = Just ghrId,
           isForwardRequest = False,
-          previousDropGeoHash = Nothing
+          previousDropGeoHash = Nothing,
+          driverTags = Just [NormalDriver, GoHomeDriverToDestination]
         }
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DriverPool.hs:640]: Why? Where? How? What?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DriverPool/Types.hs [hunk 2]
```diff
@@ -168,7 +171,8 @@ data DriverPoolWithActualDistResult = DriverPoolWithActualDistResult
     tripDistance :: Maybe Meters,
     isForwardRequest :: Bool,
     previousDropGeoHash :: Maybe Text,
-    goHomeReqId :: Maybe (Id DDGR.DriverGoHomeRequest)
+    goHomeReqId :: Maybe (Id DDGR.DriverGoHomeRequest),
+    driverTags :: Maybe [DriverTags]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DriverPool/Types.hs:175]: Name it something else

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+      Maybe Bool ->
+      m [Id Driver]
+    getPreviousBatchesDrivers mbOnRide = do
+      batches <- SDP.previouslyAttemptedDrivers searchTry.id mbOnRide
+      return $ (.driverPoolResult.driverId) <$> batches
+
+    prepareDriverPoolBatch' previousBatchesDrivers batchNum merchantOpCityId txnId isValueAddNP = withLogTag ("BatchNum - " <> show batchNum) $ do
+      radiusStep <- SDP.getPoolRadiusStep searchTry.id
+      transporterConfig <- SCTC.findByMerchantOpCityId merchantOpCityId Nothing >>= fromMaybeM (TransporterConfigDoesNotExist merchantOpCityId.getId)
+      intelligentPoolConfig <- CDIP.findByMerchantOpCityId merchantOpCityId (Just ((TransactionId . Id) txnId)) >>= fromMaybeM (InternalError "Intelligent Pool Config not found")
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs:110]: Fetch when required.

**Review by hkmangla — CHANGES_REQUESTED** [review]
Get it reviewed from @piyushKumar-1

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DriverPool.hs (hunk 1 above)
Comment (enigma-commits) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DriverPool.hs:640]: so this function is basically returning all the goHome Drivers which are going towards the destination so assigned these tags.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs (hunk 3 above)
Comment (enigma-commits) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs:113]: But we can think of optimisations also now only.

Comment (hkmangla) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs:113]: Yup. Lets first fully move to tagged pooling and then think of it.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+      blockListedDriversForSearch <- Redis.withCrossAppRedis $ Redis.getList (mkBlockListedDriversKey searchReq.id)
+      blockListedDriversForRider <- maybe (pure []) (Redis.withCrossAppRedis . Redis.getList . mkBlockListedDriversForRiderKey) searchReq.riderId
+      let blockListedDrivers = blockListedDriversForSearch <> blockListedDriversForRider
+          blockListedAndAlreadyAttemptedDrivers = blockListedDrivers <> previousBatchesDrivers
+          onlyNewAndFilteredDrivers = filter (\dp -> dp.driverPoolResult.driverId `notElem` blockListedAndAlreadyAttemptedDrivers) allDriversNotOnRide'
+          onlyNonBlockedDrivers = filter (\dp -> dp.driverPoolResult.driverId `notElem` blockListedDrivers) allDriversNotOnRide'
+
+      (driverPoolNotOnRide, driverPoolOnRide) <- do
+        case batchNum of
+          -1 -> do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs:124]: Wasn't it decided to be 0?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+                        { driverTags =
+                            insertInObject
+                              (((driverTags :: DriverPoolResult -> Value) . (driverPoolResult)) dp)
+                              [driverTag]
+                        }
+                  }
+              else dp
+        )
+
+insertInObject :: Value -> [DriverPoolTags] -> Value
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs:492]: ???

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+
+      (driverPoolNotOnRide, driverPoolOnRide) <- do
+        case batchNum of
+          -1 -> do
+            gateTaggedDrivers <- assignDriverGateTags searchReq onlyNewAndFilteredDrivers
+            goHomeTaggedDrivers <- assignDriverGoHomeTags gateTaggedDrivers searchReq searchTry tripQuoteDetails driverPoolCfg merchant goHomeConfig merchantOpCityId isValueAddNP transporterConfig
+            driverPoolOnRide <- getDriverPoolOnRide merchantOpCityId transporterConfig intelligentPoolConfig radiusStep blockListedDrivers NormalPool
+            return (goHomeTaggedDrivers, driverPoolOnRide)
+          _ -> do
+            allNearbyNonGoHomeDrivers <- filterM (\dpr -> (CQDGR.getDriverGoHomeRequestInfo dpr.driverPoolResult.driverId merchantOpCityId (Just goHomeConfig)) <&> (/= Just DDGR.ACTIVE) . (.status)) onlyNewAndFilteredDrivers
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs:130]: Why this check ?

**Review by hkmangla — CHANGES_REQUESTED** [review]

**Review by hkmangla — APPROVED** [review]

**Review by hkmangla — APPROVED** [review]

**Pull request merged**